<a href="https://colab.research.google.com/github/lblogan14/web_scraping_with_python/blob/master/ch13_images_text.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdocs.test%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.photos.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


In [5]:
%cd /content/drive/My' 'Drive/Colab' 'Notebooks/Web_Scraping_with_Python

/content/drive/My Drive/Colab Notebooks/Web_Scraping_with_Python


#Overview of Libraries
Translating images into text is called *optical character recognition*, or *OCR*. 

##Pillow
Pillow allows you to easily import and manipulate images with a variety of filters, masks, and even pixel-specific transformations. In this chapter, Pillow is used to perform preprocessing on images to make them more machine readable.

##Tesseract
Tesseract is an OCR library. This chapter uses both the command-line program *Tesseract* along with its third-party Python wrapper *pytesseract*.

###Installing Tesseract


In [10]:
!pip3 install pytesseract

In [16]:
!apt-get install tesseract-ocr

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following package was automatically installed and is no longer required:
  libnvidia-common-410
Use 'apt autoremove' to remove it.
The following additional packages will be installed:
  tesseract-ocr-eng tesseract-ocr-osd
The following NEW packages will be installed:
  tesseract-ocr tesseract-ocr-eng tesseract-ocr-osd
0 upgraded, 3 newly installed, 0 to remove and 4 not upgraded.
Need to get 4,795 kB of archives.
After this operation, 15.8 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 tesseract-ocr-eng all 4.00~git24-0e00fe6-1.2 [1,588 kB]
Get:2 http://archive.ubuntu.com/ubuntu bionic/universe amd64 tesseract-ocr-osd all 4.00~git24-0e00fe6-1.2 [2,989 kB]
Get:3 http://archive.ubuntu.com/ubuntu bionic/universe amd64 tesseract-ocr amd64 4.00~git2288-10f4998a-2 [218 kB]
Fetched 4,795 kB in 1s (4,331 kB/s)
Selecting previously unselecte

In [6]:
%cd ./img/ch13/

/content/drive/My Drive/Colab Notebooks/Web_Scraping_with_Python/img/ch13


In [0]:
from PIL import Image
import pytesseract

In [19]:
print(pytesseract.image_to_string(Image.open('textOriginal.png')))

This is some text, written in Arial, that will be read by
Tesseract. Here are some symbols: !|@#$%&*()


To point pytesseract to the location by including this line:

```
pytesseract.pytesseract.tesseract_cmd = '/path/to/tesseract'
```

Pytesseract can also estimate box files (pixel locations for the boundaries of each character):

In [21]:
print(pytesseract.image_to_boxes(Image.open('textOriginal.png')))

T 23 60 29 77 0
h 29 60 36 77 0
i 40 60 49 77 0
s 52 60 66 77 0
i 76 60 78 77 0
s 81 60 91 73 0
s 99 60 109 73 0
o 113 60 122 73 0
m 125 60 134 73 0
e 134 60 142 73 0
t 144 60 156 73 0
e 166 60 170 77 0
x 176 60 182 73 0
t 188 60 195 73 0
, 196 56 206 77 0
w 215 60 225 73 0
r 225 60 233 73 0
i 235 60 240 73 0
t 243 60 252 77 0
t 253 60 259 77 0
e 260 60 269 73 0
n 275 60 281 73 0
i 295 60 297 77 0
n 300 60 309 73 0
A 317 60 325 77 0
r 325 60 333 77 0
i 335 60 340 73 0
a 343 60 353 77 0
l 353 60 358 73 0
, 361 56 369 77 0
t 378 60 384 77 0
h 388 60 395 77 0
a 399 60 408 73 0
t 412 60 418 77 0
w 425 60 430 73 0
i 430 60 441 73 0
l 444 60 451 77 0
l 455 60 457 77 0
b 467 60 477 77 0
e 480 60 491 73 0
r 500 60 506 73 0
e 511 60 519 73 0
a 521 60 532 73 0
d 534 60 545 77 0
b 555 60 564 77 0
y 567 55 578 73 0
T 23 31 29 48 0
e 29 31 41 48 0
s 41 31 46 44 0
s 52 31 59 44 0
e 64 31 71 44 0
r 76 31 84 44 0
a 87 31 93 44 0
c 99 31 105 44 0
t 109 31 118 44 0
. 121 31 130 48 0
H 140 31 153 48 0
e 

It can also return a complete output of all data, such as confidence scores, page and
line numbers, box data, as well as other information:

In [22]:
print(pytesseract.image_to_data(Image.open('textOriginal.png')))

level	page_num	block_num	par_num	line_num	word_num	left	top	width	height	conf	text
1	1	0	0	0	0	0	0	600	103	-1	
2	1	1	0	0	0	23	26	555	51	-1	
3	1	1	1	0	0	23	26	555	51	-1	
4	1	1	1	1	0	23	26	555	22	-1	
5	1	1	1	1	1	23	26	44	17	96	This
5	1	1	1	1	2	76	26	15	17	96	is
5	1	1	1	1	3	99	30	43	13	96	some
5	1	1	1	1	4	144	26	62	21	96	text,
5	1	1	1	1	5	215	26	70	17	96	written
5	1	1	1	1	6	295	26	15	17	96	in
5	1	1	1	1	7	317	26	52	21	96	Arial,
5	1	1	1	1	8	378	26	40	17	96	that
5	1	1	1	1	9	425	26	32	17	96	will
5	1	1	1	1	10	467	26	24	17	96	be
5	1	1	1	1	11	500	26	45	17	96	read
5	1	1	1	1	12	555	26	23	22	96	by
4	1	1	1	2	0	23	54	513	23	-1	
5	1	1	1	2	1	23	55	107	17	96	Tesseract.
5	1	1	1	2	2	140	55	49	17	96	Here
5	1	1	1	2	3	198	59	32	13	96	are
5	1	1	1	2	4	239	59	57	13	96	some
5	1	1	1	2	5	304	55	92	22	90	symbols:
5	1	1	1	2	6	407	54	129	23	64	!|@#$%&*()


The default output for these last two files is as space- or tab-delimited string files, but it can also be used to get output as dictionaries or byte strings:

In [0]:
from pytesseract import Output

In [25]:
print(pytesseract.image_to_data(Image.open('textOriginal.png'),
                                output_type = Output.DICT))

{'level': [1, 2, 3, 4, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 4, 5, 5, 5, 5, 5, 5], 'page_num': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], 'block_num': [0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], 'par_num': [0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], 'line_num': [0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2], 'word_num': [0, 0, 0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 0, 1, 2, 3, 4, 5, 6], 'left': [0, 23, 23, 23, 23, 76, 99, 144, 215, 295, 317, 378, 425, 467, 500, 555, 23, 23, 140, 198, 239, 304, 407], 'top': [0, 26, 26, 26, 26, 26, 30, 26, 26, 26, 26, 26, 26, 26, 26, 26, 54, 55, 55, 59, 59, 55, 54], 'width': [600, 555, 555, 555, 44, 15, 43, 62, 70, 15, 52, 40, 32, 24, 45, 23, 513, 107, 49, 32, 57, 92, 129], 'height': [103, 51, 51, 22, 17, 17, 13, 21, 17, 17, 21, 17, 17, 17, 17, 22, 23, 17, 17, 13, 13, 22, 23], 'conf': ['-1', '-1', '-1', '-1', 96, 96, 96, 96, 96, 96, 96, 96, 96, 96

In [26]:
print(pytesseract.image_to_string(Image.open('textOriginal.png'),
                                  output_type = Output.BYTES))

b'This is some text, written in Arial, that will be read by\nTesseract. Here are some symbols: !|@#$%&*()\n\x0c'


##NumPy
Standard Python library

#Processing Well-Formatted Text
To run Tesseract from the command line to read this file and write the results to a text file:

In [29]:
!tesseract textOriginal.png textoutput | cat textoutput.txt

This is some text, written in Arial, that will be read by
Tesseract. Here are some symbols: !|@#$%&*()
Tesseract Open Source OCR Engine v4.0.0-beta.1 with Leptonica


The result is almost accurate...
![](https://github.com/lblogan14/web_scraping_with_python/blob/master/img/ch13/textOriginal.png?raw=true)

For a blurred image below, the result could be worse:
![](https://github.com/lblogan14/web_scraping_with_python/blob/master/img/ch13/textBad.png?raw=true)

In [30]:
!tesseract textBad.png textoutput | cat textoutput.txt

This is some text, written in Arial, that will be read by
Tesseract. Here are some symbols: !|@#$%&*()
Tesseract Open Source OCR Engine v4.0.0-beta.1 with Leptonica


To clean the images, try the following script:

In [0]:
from PIL import Image
import pytesseract

In [0]:
def cleanFile(filePath, newFilePath):
  image = Image.open(filePath)
  
  #Set a threshold value for the image, and save
  image = image.point(lambda x: 0 if x<143 else 255)
  image.save(newFilePath)
  return image

In [0]:
image = cleanFile('./textBad.png', './textCleaned.png')

In [36]:
print(pytesseract.image_to_string(image))

This 1s some text, written in Arial, that will be read by
Tesseract Here are some symbols: |@#$%*&"(}


##Adjusting Images Automatically

To run Tesseract against a range of images adjusted to different values and algorithmically choose the one with the best result, as measured by some combination of the number of characters and/or strings Tesseract is able to read, and the “confidence” that it reads those characters with.

In [0]:
import pytesseract
from pytesseract import Output
from PIL import Image
import numpy as np

In [0]:
def cleanFile(filePath, threshold):
  image = Image.open(filePath)
  
  #Set a threshold value for the image, and save
  image = image.point(lambda x: 0 if x<threshold else 255)
  return image

In [0]:
def getConfidence(image):
  data = pytesseract.image_to_data(image, output_type=Output.DICT)
  text = data['text']
  confidences = []
  numChars = []
  
  for i in range(len(text)):
    if list(map(int,data['conf']))[i] > -1:
      confidences.append(data['conf'][i])
      numChars.append(len(text[i]))
      
  return np.average(confidences, weights=numChars), sum(numChars)

In [61]:
filePath = './textBad.png'

start = 80
step = 5
end = 200

for threshold in range(start, end, step):
  image = cleanFile(filePath, threshold)
  scores = getConfidence(image)
  print('threshold: {}, confidence: {}, numChars {}'.format(str(threshold), 
                                                            str(scores[0]), 
                                                            str(scores[1])))

threshold: 80, confidence: 36.10526315789474, numChars 19
threshold: 85, confidence: 62.333333333333336, numChars 24
threshold: 90, confidence: 49.25, numChars 36
threshold: 95, confidence: 39.7, numChars 40
threshold: 100, confidence: 49.98360655737705, numChars 61
threshold: 105, confidence: 48.83098591549296, numChars 71
threshold: 110, confidence: 73.87654320987654, numChars 81
threshold: 115, confidence: 73.6219512195122, numChars 82
threshold: 120, confidence: 77.06024096385542, numChars 83
threshold: 125, confidence: 83.61904761904762, numChars 84
threshold: 130, confidence: 75.04761904761905, numChars 84
threshold: 135, confidence: 82.42168674698796, numChars 83
threshold: 140, confidence: 81.76190476190476, numChars 84
threshold: 145, confidence: 85.51190476190476, numChars 84
threshold: 150, confidence: 82.3132530120482, numChars 83
threshold: 155, confidence: 87.9873417721519, numChars 79
threshold: 160, confidence: 85.8875, numChars 80
threshold: 165, confidence: 86.1, numC

Threshold of 145 gives the maximum number of recognized characters and the highest confidence for those found characters.

##Scraping Text from Images on Websites
The following script navigates to the large-print edition of Tolstoy's The Death of Ivan Ilyich, opens the reader, collects image URLs, and then
systematically downloads, reads, and prints the text from each one.

Note that this code depends on a live Amazon listing as well as several architectural
features of the Amazon website to run correctly.

In [63]:
!pip3 install selenium

     |████████████████████████████████| 911kB 2.8MB/s 


In [0]:
import time
from urllib.request import urlretrieve
from PIL import Image
import pytesseract
from selenium import webdriver
from PIL import Image

In [0]:
#Create new Selenium driver
driver = webdriver.Chrome(executable_path='drivers/chromedriver/chromedriver')

driver.get('https://www.amazon.com/Death-Ivan_Ilyich-Nikolayevich-Tolstoy/dp/1427027277')
time.sleep(2)

#Click on the book preview button
driver.find_element_by_id('imgBlkFront').click()
imageList = []

#Wait for the page to load
time.sleep(5)

while 'pointer' in driver.find_element_by_id('siteReaderRightPageTurner').get_attribute('style'):
  #While the right arrow is available for clicking, turn through pages
  driver.find_element_by_id('siteReaderRightPageTurner').click()
  time.sleep(2)
  #Get any new pages that have loaded (multiple pages can load at once,
  #but duplicates will not be added to a set)
  pages = driver.find_elements_by_xpath('//div[@class=\'pageImage\']/div/img')
  if not len(pages):
    print('No pages found')
  for page in pages:
    image = page.get_attribute('src')
    print('Found image: {}'.format(image))
    if image not in imageList:
      urlretrieve(image, 'page.jpg')
      imageList.append(image)
      print(pytesseract.image_to_string(Image.open('page.jpg')))
      
driver.quit()